File: twidere/src/main/java/org/mariotaku/twidere/util/ReadStateManager.java
/*
 * 				Twidere - Twitter client for Android
 * 
 *  Copyright (C) 2012-2014 Mariotaku Lee <mariotaku.lee@gmail.com>
 * 
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 * 
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 * 
 *  You should have received a copy of the GNU General Public License
 *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package org.mariotaku.twidere.util;

import android.content.Context;
import android.content.SharedPreferences;
import android.content.SharedPreferences.OnSharedPreferenceChangeListener;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.text.TextUtils;

import org.mariotaku.twidere.annotation.CustomTabType;
import org.mariotaku.twidere.annotation.NotificationType;
import org.mariotaku.twidere.annotation.ReadPositionTag;
import org.mariotaku.twidere.model.StringLongPair;
import org.mariotaku.twidere.util.collection.CompactHashSet;

import java.util.Set;

import static org.mariotaku.twidere.TwidereConstants.TIMELINE_POSITIONS_PREFERENCES_NAME;

public class ReadStateManager {

    private final SharedPreferencesWrapper mPreferences;

    public ReadStateManager(final Context context) {
        mPreferences = SharedPreferencesWrapper.getInstance(context,
                TIMELINE_POSITIONS_PREFERENCES_NAME, Context.MODE_PRIVATE);
    }

    public long getPosition(final String key) {
        if (TextUtils.isEmpty(key)) return -1;
        return mPreferences.getLong(key, -1);
    }

    @NonNull
    public StringLongPair[] getPositionPairs(final String key) {
        if (TextUtils.isEmpty(key)) return new StringLongPair[0];
        final Set<String> set = mPreferences.getStringSet(key, null);
        if (set == null) return new StringLongPair[0];
        final StringLongPair[] pairs = new StringLongPair[set.size()];
        int count = 0;
        for (String entry : set.toArray(new String[set.size()])) {
            try {
                pairs[count++] = StringLongPair.valueOf(entry);
            } catch (NumberFormatException e) {
                return new StringLongPair[0];
            }
        }
        return pairs;
    }

    public long getPosition(final String key, final String keyId) {
        if (TextUtils.isEmpty(key)) return -1;
        final Set<String> set = mPreferences.getStringSet(key, null);
        if (set == null) return -1;
        final String prefix = keyId + ":";
        for (String entry : set) {
            if (entry.startsWith(prefix)) return StringLongPair.valueOf(entry).getValue();
        }
        return -1;
    }

    public void registerOnSharedPreferenceChangeListener(OnSharedPreferenceChangeListener listener) {
        mPreferences.registerOnSharedPreferenceChangeListener(listener);
    }

    public boolean setPosition(String key, String keyId, long position) {
        return setPosition(key, keyId, position, false);
    }

    public void unregisterOnSharedPreferenceChangeListener(OnSharedPreferenceChangeListener listener) {
        mPreferences.unregisterOnSharedPreferenceChangeListener(listener);
    }


    public boolean setPosition(final String key, final String keyId, final long position, boolean acceptOlder) {
        if (TextUtils.isEmpty(key)) return false;
        Set<String> set = mPreferences.getStringSet(key, null);
        if (set == null) {
            set = new CompactHashSet<>();
        }
        String keyValue = null;
        final String prefix = keyId + ":";
        for (String entry : set) {
            if (entry.startsWith(prefix)) {
                keyValue = entry;
                break;
            }
        }
        final StringLongPair pair;
        if (keyValue != null) {
            // Found value
            pair = StringLongPair.valueOf(keyValue);
            if (!acceptOlder && pair.getValue() > position) return false;
            set.remove(keyValue);
            pair.setValue(position);
        } else {
            pair = new StringLongPair(keyId, position);
        }
        set.add(pair.toString());
        final SharedPreferences.Editor editor = mPreferences.edit();
        editor.putStringSet(key, set);
        editor.apply();
        return true;
    }


    public boolean setPositionPairs(final String key, @Nullable final StringLongPair[] pairs) {
        if (TextUtils.isEmpty(key)) return false;
        final SharedPreferences.Editor editor = mPreferences.edit();
        if (pairs == null) {
            editor.remove(key);
        } else {
            final Set<String> set = new CompactHashSet<>();
            for (StringLongPair pair : pairs) {
                set.add(pair.toString());
            }
            editor.putStringSet(key, set);
        }
        editor.apply();
        return true;
    }

    public boolean setPosition(final String key, final long position) {
        return setPosition(key, position, false);
    }

    public boolean setPosition(final String key, final long position, boolean acceptOlder) {
        if (TextUtils.isEmpty(key) || !acceptOlder && getPosition(key) >= position) return false;
        final SharedPreferences.Editor editor = mPreferences.edit();
        editor.putLong(key, position);
        editor.apply();
        return true;
    }

    public interface OnReadStateChangeListener {
        void onReadStateChanged();
    }

    @Nullable
    @ReadPositionTag
    public static String getReadPositionTagForNotificationType(@NotificationType String notificationType) {
        if (notificationType == null) return null;
        switch (notificationType) {
            case NotificationType.HOME_TIMELINE: {
                return ReadPositionTag.HOME_TIMELINE;
            }
            case NotificationType.DIRECT_MESSAGES: {
                return ReadPositionTag.DIRECT_MESSAGES;
            }
            case NotificationType.INTERACTIONS: {
                return ReadPositionTag.ACTIVITIES_ABOUT_ME;
            }
        }
        return null;
    }

    @Nullable
    @ReadPositionTag
    public static String getReadPositionTagForTabType(@CustomTabType String tabType) {
        if (tabType == null) return null;
        switch (tabType) {
            case CustomTabType.HOME_TIMELINE: {
                return ReadPositionTag.HOME_TIMELINE;
            }
            case CustomTabType.NOTIFICATIONS_TIMELINE: {
                return ReadPositionTag.ACTIVITIES_ABOUT_ME;
            }
            case CustomTabType.DIRECT_MESSAGES: {
                return ReadPositionTag.DIRECT_MESSAGES;
            }
        }
        return null;
    }

}

File: twidere/src/main/kotlin/org/mariotaku/twidere/util/ReadStateManager.kt
/*
 * 				Twidere - Twitter client for Android
 * 
 *  Copyright (C) 2012-2014 Mariotaku Lee <mariotaku.lee@gmail.com>
 * 
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 * 
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 * 
 *  You should have received a copy of the GNU General Public License
 *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package org.mariotaku.twidere.util

import android.content.Context
import android.content.SharedPreferences.OnSharedPreferenceChangeListener
import android.text.TextUtils
import org.mariotaku.twidere.TwidereConstants.TIMELINE_POSITIONS_PREFERENCES_NAME
import org.mariotaku.twidere.annotation.CustomTabType
import org.mariotaku.twidere.annotation.NotificationType
import org.mariotaku.twidere.annotation.ReadPositionTag

class ReadStateManager(context: Context) {

    private val preferences = SharedPreferencesWrapper.getInstance(context,
            TIMELINE_POSITIONS_PREFERENCES_NAME, Context.MODE_PRIVATE)

    fun getPosition(key: String): Long {
        if (TextUtils.isEmpty(key)) return -1
        return preferences.getLong(key, -1)
    }

    fun registerOnSharedPreferenceChangeListener(listener: OnSharedPreferenceChangeListener) {
        preferences.registerOnSharedPreferenceChangeListener(listener)
    }

    fun unregisterOnSharedPreferenceChangeListener(listener: OnSharedPreferenceChangeListener) {
        preferences.unregisterOnSharedPreferenceChangeListener(listener)
    }

    fun setPosition(key: String, position: Long, acceptOlder: Boolean = false): Boolean {
        if (TextUtils.isEmpty(key) || !acceptOlder && getPosition(key) >= position) return false
        val editor = preferences.edit()
        editor.putLong(key, position)
        editor.apply()
        return true
    }

    companion object {

        @ReadPositionTag
        fun getReadPositionTagForNotificationType(@NotificationType notificationType: String?): String? {
            if (notificationType == null) return null
            when (notificationType) {
                NotificationType.HOME_TIMELINE -> {
                    return ReadPositionTag.HOME_TIMELINE
                }
                NotificationType.DIRECT_MESSAGES -> {
                    return ReadPositionTag.DIRECT_MESSAGES
                }
                NotificationType.INTERACTIONS -> {
                    return ReadPositionTag.ACTIVITIES_ABOUT_ME
                }
            }
            return null
        }

        @ReadPositionTag
        fun getReadPositionTagForTabType(@CustomTabType tabType: String?): String? {
            if (tabType == null) return null
            when (tabType) {
                CustomTabType.HOME_TIMELINE -> {
                    return ReadPositionTag.HOME_TIMELINE
                }
                CustomTabType.NOTIFICATIONS_TIMELINE -> {
                    return ReadPositionTag.ACTIVITIES_ABOUT_ME
                }
                CustomTabType.DIRECT_MESSAGES -> {
                    return ReadPositionTag.DIRECT_MESSAGES
                }
            }
            return null
        }
    }

}
